toolchain/check/handle_binding_pattern.cpp

```diff
@@ -35,28 +39,21 @@ auto HandleBindingPattern(Context& context, Parse::NodeId parse_node) -> bool {
   auto type_node_copy = type_node;
   auto cast_type_id = ExprAsType(context, type_node, parsed_type_id);
 
-  // A `self` binding doesn't have a name.
-  if (auto self_node =
-          context.node_stack()
-              .PopForSoloParseNodeIf<Parse::NodeKind::SelfValueName>()) {
-    if (context.parse_tree().node_kind(context.node_stack().PeekParseNode()) !=
-        Parse::NodeKind::ImplicitParamListStart) {
-      CARBON_DIAGNOSTIC(
-          SelfOutsideImplicitParamList, Error,
-          "`self` can only be declared in an implicit parameter list.");
-      context.emitter().Emit(parse_node, SelfOutsideImplicitParamList);
-    }
-    context.AddInstAndPush(
-        parse_node, SemIR::SelfParam{*self_node, cast_type_id,
-                                     /*is_addr_self=*/SemIR::BoolValue::False});
-    return true;
-  }
-
   // TODO: Handle `_` bindings.
 
   // Every other kind of pattern binding has a name.
   auto [name_node, name_id] = context.node_stack().PopNameWithParseNode();
 
+  // A `self` binding can only appear in an implicit parameter list.
+  if (name_id == SemIR::NameId::SelfValue &&
+      context.parse_tree().node_kind(context.node_stack().PeekParseNode()) !=
+          Parse::NodeKind::ImplicitParamListStart) {
```

Comment:
```suggestion
      context.PeekIs<Parse::NodeKind::ImplicitParamListStart>()) {
```

Looks pre-existing, but feels worth cleaning up while you're changing the code. Should probably do a pass trying to clean up things like this.